Clipping widgets

drug007 opened the pull request:
More advanced and simpler mechanism of widget clipping.

drug007 committed:
Rollback part of previous changes.

kiith-sa commented:
Just from a reading review: 

Is there really need for a string mixin as opposed to a function for `checkInWidgetScreenXXX`? 

It seems to me that something `bool isWidgetInScreen(size_t spacing)()` (or `isWidgetInArea`?) combined with `if(isWidgetInScreen!WIDGET_Y_SPACING) { return false;}`/ `if(isWidgetInScreen!WIDGET_Y_SPACING) { return;}` would still work, would be more readable to D newcomers and more hygienic (no stealth-inserted `minx`/`miny` local variables); the overhead should be minimal or zero, depending on optimizations (and certainly negligible compared to other `imgui` code)

I'd prefer to only resort to string mixins for hardcore code generation where even templates aren't sufficient.

(Also, I think `g_state.widgetY -= WIDGET_Y_SPACING;` should be loudly documented in a DDoc comment as it is a side effect; it's true that `imgui` contains tons of undocumented side effects but I'd like to not add many more so the code could gradually become more, not less, readable.

drug007 commented:
I agree with you on that. It'll take some time (couple of days I think) for me to complete PR.

drug007 committed:
More advanced and simple mechanism of widget clipping.

We just check if a widget is within current scroll area and if not do not add any drawing command (but keep tracking Y position of the widget to provide scrolling)

drug007 commented:
I rewrite it more simpler, review please again - changing Y position of the current widget is performed only in rendering functions (due to `scope(exit)`).
`isWidgetInScreen` have no side effect anymore.

kiith-sa commented:
It seems OK (didn't try to build yet either), but this seems needlessly verbose:

```
+    if(g_state.widgetY < miny || g_state.widgetY > maxy)
+        return false;
+    else
+        return true; 
```

could be just:

```
return !(g_state.widgetY < miny || g_state.widgetY > maxy);
```

drug007 commented:
I agree with you again.
